### PR TITLE
dfdl processor example that integrates with pubsub and firestore

### PR DESCRIPTION
**Pull request by carolinahdez**
This module is a example how to process a binary using a DFDL definition. The DFDL definitions are stored in a Firestore database.

The application send a request with the binary to process to a pubsub topic.

The processor service subscribes to the topic, processes every message, applies the definition and publishes the json result to a topic in pubsub.

### CODE REVIEW

**Comment by mbeckerle** [comment]
Thank you for this submission. Very cool. I will look at this code soon. 

First thing however, is would you like to offer this under the terms of the Apache Public License v2.0? If so adding a LICENSE.txt file to that effect would be good. 

(Can copy from here: https://github.com/OpenDFDL/examples/blob/master/pairsTransform/LICENSE.txt)

I noticed some other examples I created are missing this License as well, so I'll be adding those.

**Comment by mbeckerle** [comment]
Ah nevermind. I neglected to look in the repo root dir, which already has a LICENSE file.

**Review by stevedlawrence — APPROVED** [review]
Looks very cool! :+1: 

I'm not too familiar with some of these technologies, so can't comment too much on those parts--but just a couple minor comments on the DFDL/Daffodil aspects.

File: dfdl_example_pubsub_firestore/src/main/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (mbeckerle) [inline dfdl_example_pubsub_firestore/src/main/resources/application.properties:3]: And of course I didn't see that this file is the license until now. :-)
We're good on license.

**Comment by mbeckerle** [comment]
> Thank you for this submission. Very cool. I will look at this code soon.
> 
> First thing however, is would you like to offer this under the terms of the Apache Public License v2.0? If so adding a LICENSE.txt file to that effect would be good.
> 
> (Can copy from here: https://github.com/OpenDFDL/examples/blob/master/pairsTransform/LICENSE.txt)
> 
> I noticed some other examples I created are missing this License as well, so I'll be adding those.

Sheesh. I can't believe I nagged you about the license stuff when application.properties has the APL v2 and every single source file also is marked that way. I guess I should read ahead a bit more before commenting. Sorry for the annoyance.

File: dfdl_example_pubsub_firestore/src/main/java/com/example/dfdl/DfdlService.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+ * Transforms messages using the {@link Daffodil} DFDL processor.
+ *
+ * <p>The {@link Daffodil} object is a factory object to create a {@link Compiler}. The Compiler
+ * provides a method to compile a provided DFDL schema into a {@link ProcessorFactory}, which
+ * creates a {@link DataProcessor}.
+ */
+@Service
+public class DfdlService {
+
+  public String convertDataMessage(String message, DfdlDef dfdl) throws IOException {
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply dfdl_example_pubsub_firestore/src/main/java/com/example/dfdl/DfdlService.java:45]: Yeah, perhaps a future enhancement, but you seem to have the perfect place to store the compiled DFDL schema which is along with the text of that schema in the DFDLDefRepository.  I wasn't exactly sure when schemas get inserted into that repository, but getting any schema compilation errors/warnings at that time would seem to be worthwhile vs. waiting until the schema is used to parse data.  

For now the only thing I'd suggest is put a comment in the code here like 

`// todo: save compiled DFDL schema so we don't compile it for every parse.`

**Review by mbeckerle — APPROVED** [review]
+1 I think minor cleanups per steve lawrence's comments to the example DFDL schema would be nice, but really nothing has to be fixed.  I am looking forward to trying this out.

File: dfdl_example_pubsub_firestore/examples/bynary_example.dfdl.xsd [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  <xs:element name="binary_example">
+    <xs:complexType>
+      <xs:sequence>
+        <xs:element name="w" type="xs:int">
+          <xs:annotation>
+            <xs:appinfo source="http://www.ogf.org/dfdl/">
+              <dfdl:element binaryNumberRep="binary"
+                  byteOrder="bigEndian"
+                  lengthKind="implicit"
+                  representation="binary"/>
```
[the remark this answers is not in the document]
Comment (carolinahdez) [reply dfdl_example_pubsub_firestore/examples/bynary_example.dfdl.xsd]: Done

File: dfdl_example_pubsub_firestore/pom.xml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      <artifactId>spring-boot-starter-test</artifactId>
+      <groupId>org.springframework.boot</groupId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <artifactId>spring-boot-starter-web</artifactId>
+      <groupId>org.springframework.boot</groupId>
+    </dependency>
+    <dependency>
+      <artifactId>daffodil-core_2.12</artifactId>
```
[the remark this answers is not in the document]
Comment (carolinahdez) [reply dfdl_example_pubsub_firestore/pom.xml]: Done

File: dfdl_example_pubsub_firestore/src/main/java/com/example/dfdl/DfdlService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (carolinahdez) [reply dfdl_example_pubsub_firestore/src/main/java/com/example/dfdl/DfdlService.java]: Added comment about good practice to implement this in the cloud and a TODO

File: dfdl_example_pubsub_firestore/README.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
+# Data Format Description Language ([DFDL](https://en.wikipedia.org/wiki/Data_Format_Description_Language)) Processor Example
+This module is a example how to process a binary using a DFDL definition.
```
[the remark this answers is not in the document]
Comment (carolinahdez) [reply dfdl_example_pubsub_firestore/README.md:2]: Will do in a follow up CL

**Comment by carolinahdez** [comment]
All the suggestions have been addressed except for renaming the directory which will be done in a follow up PR. Thanks.

**Pull request merged**